geode-core/src/test/java/org/apache/geode/management/JMXMBeanReconnectDUnitTest.java

```diff
@@ -0,0 +1,314 @@
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import org.apache.geode.cache.Cache;
+import org.apache.geode.distributed.ConfigurationProperties;
+import org.apache.geode.internal.AvailablePortHelper;
+import org.apache.geode.management.internal.SystemManagementService;
+import org.apache.geode.test.dunit.rules.ClusterStartupRule;
+import org.apache.geode.test.dunit.rules.MemberVM;
+import org.apache.geode.test.junit.categories.DistributedTest;
+import org.apache.geode.test.junit.categories.JMXTest;
+import org.apache.geode.test.junit.rules.GfshCommandRule;
+import org.apache.geode.test.junit.rules.MBeanServerConnectionRule;
+
+@Category({DistributedTest.class, JMXTest.class})
+
+public class JMXMBeanReconnectDUnitTest {
+  private MemberVM locator1, locator2, server1, server2;
+
+  private int jmxPort1;
+  private String locator1Name = "locator-one";
+
+  private int jmxPort2;
+  private String locator2Name = "locator-two";
+
+  private String regionName1 = "test-region-1";
+
+  @Rule
+  public ClusterStartupRule lsRule = new ClusterStartupRule();
+
+  @Rule
+  public GfshCommandRule gfsh = new GfshCommandRule();
+
+  @Rule
+  public MBeanServerConnectionRule jmxConnectionRule = new MBeanServerConnectionRule();
+
+  private int locator1VMIndex = 0;
+  private int locator2VMIndex = 1;
+  private int server1VMIndex = 2;
+  private int server2VMIndex = 3;
+  private int restoredMemberVMIndex = 4;
+
+  @Before
+  public void before() throws Exception {
+    int[] ports = AvailablePortHelper.getRandomAvailableTCPPorts(2);
+    jmxPort1 = ports[0];
+    jmxPort2 = ports[1];
+
+    // Start a locator1, a server1, another locator1, and a region
+    locator1 = lsRule.startLocatorVM(locator1VMIndex, getLocator1Properties());
+    locator2 = lsRule.startLocatorVM(locator2VMIndex, getLocator2Properties(), locator1.getPort());
+
+    server1 = lsRule.startServerVM(server1VMIndex, locator1.getPort());
+    server2 = lsRule.startServerVM(server2VMIndex, locator1.getPort());
+
+    gfsh.connectAndVerify(locator1);
+    gfsh.executeAndAssertThat("create region --type=REPLICATE --name=/" + regionName1)
+        .statusIsSuccess();
+
+    // Avoid a minor race condition if locators haven't come to terms yet.
+    Awaitility.waitAtMost(1, TimeUnit.MINUTES)
+        .until(
+            () -> getAllFederatedGemfireBeanObjectNamesFromRemoteBeanServer(locator1.getJmxPort())
+                .size() == getAllFederatedGemfireBeanObjectNamesFromRemoteBeanServer(
+                    locator2.getJmxPort())
+                        .size());
+    locator1.waitTillRegionsAreReadyOnServers("/" + regionName1, 2);
+  }
+
+  @Test
+  public void testLocalBeans_MaintainServerAndCrashLocator() {
+    List<String> initialServerBeans = server1.invoke(
+        JMXMBeanReconnectDUnitTest::getLocalCanonicalBeanNames);
+
+    int portOfCrashedMember = locator1.getPort();
+    lsRule.stopVM(locator1VMIndex);
```

Comment:
If the method name is accurate and you actually do wish to crash a system then `stopVM` is not the right call. (It will end up doing a graceful shutdown). Alternatively you can use `MembershipManagerHelper.crashDistributedSystem(InternalDistributedSystem.getConnectedInstance())`. I've used this in `ClusterConfigLocatorRestartDUnitTest`. There may also be other suitable methods like `playDead` for you to try.

Reply:
w/ @dhemery: After playing around with different ways of crashing members we decided to keep using stopVM for now and to file a ticket to add more functionality to the clusterStarterRule. In the interest of doing things right, we don't want to expand that rule in this ticket.